sphinx/environment/__init__.py

```diff
@@ -819,19 +819,28 @@ def _traverse_toctree(
                 traversed.add(sub_docname)
 
 
-def _check_toc_parents(toctree_includes: dict[str, list[str]]) -> None:
+def _check_toc_parents(toctree_includes: dict[str, list[str]], root_doc: str) -> None:
+    """Checks if document is referenced in multiple toctrees.
+    Based on the current implementation of `global_toctree_for_doc`,
+    it considers only the descendants of root_doc and not the whole graph.
+    """
     toc_parents: dict[str, list[str]] = {}
-    for parent, children in toctree_includes.items():
-        for child in children:
-            toc_parents.setdefault(child, []).append(parent)
 
+    def _find_toc_parents_dfs(node: str) -> None:
+        for child in toctree_includes.get(node, []):
+            toc_parents.setdefault(child, []).append(node)
+            is_child_already_visited = len(toc_parents[child]) > 1
+            if not is_child_already_visited:
+                _find_toc_parents_dfs(child)
+
+    _find_toc_parents_dfs(root_doc)
     for doc, parents in sorted(toc_parents.items()):
         if len(parents) > 1:
             logger.info(
                 __(
                     'document is referenced in multiple toctrees: %s, selecting: %s <- %s'
                 ),
-                parents,
+                sorted(parents),
```

Comment:
Was the sorting added here for debugging/investigation purposes?  (and should we include it with these changes?)

Reply:
The helper function uses preorder traversal which does not guarantee sorted parents as was before. Sorting is kept for consistency reasons (independent of the helper function traversal order) in the logged output, this way it is also easier to right the corresponding tests instead of dry running the traversal order and depending on the helper functions implementation. Further, it also makes it easier for the user to spot the pattern that the lexicographically greatest parent is being selected.

Reply:
My initial sense here is that I'm not too keen on the practice of modifying application code in order to make test expectations easier to write.

I do understand that it helps in this case, but I think that unit test coverage of different tree/graph structures would be more robust over time.

(apologies for taking a while to add further review commentary)

Reply:
> My initial sense here is that I'm not too keen on the practice of modifying application code  
I do not understand how the application code is modified since the function `check_toc_parents()` does not produce any side-effects other than the output. In fact I think the idea of applying `sorted` is quite the opposite.  
I'd like to clarify again that the mentioned benefits/reasons in my previous comment of having sorted parents aren't enforced by this PR, instead the parents were already sorted previously due to the node-wise traversal and the sorted nature of values in `toctree_includes`. Since the traversal order is now changed to inorder which doesn't inherently guarantee parents being collected in sorted order, `sorted` function is now applied post-traversal, to keep it consistent with the previous behavior. You could argue that guaranteeing the order of parents should come from the nested helper function itself and the outer body of `check_toc_parents()` shall not be modified, however given the recursive nature of the helper function, I feel like the current approach is much simpler.  

> I do understand that it helps in this case, but I think that unit test coverage of different tree/graph structures would be more robust over time.  
I don’t have a strong opinion on writing unittests for helper functions, in my opinion we should be testing based on functionality and not the implementation of a function which in this current case would mean that the tests should only care about consistent warning/logging and not about whatever method of traversal is used internally to achieve so. For example the helper method which used node-wise traversal previously, and now inorder traversal in this PR, should ideally NOT break the existing tests and hence having a determined order of parents regardless of the traversal algorithm helps achieve it.  

I’d like to know more about what you think of this. If you still believe that we shouldn’t guarantee sorted order of parents anymore, I’d happily remove it.

Reply:
> > My initial sense here is that I'm not too keen on the practice of modifying application code
> 
> I do not understand how the application code is modified since the function `_check_toc_parents()` does not produce any side-effects other than the output.

Logging is sometime more than a side-effect, and it can either express or hide internal application state.  Sometimes that's risky, and sometimes it's useful.  In the context of trying to improve the consistency of build output, I think it's likely to be useful to log the iteration order of the parent document names without sorting applied to them.

As I understand it, commit 8351936994df62aadcc299af3c8106893b746a14 does sort the keys of the toctree includes -- but it doesn't sort the values (the parent document list).

Reply:
Thanks for the feedback, I'll remove sorting from here then. 
<br/><br/>
> As I understand it, commit https://github.com/sphinx-doc/sphinx/commit/8351936994df62aadcc299af3c8106893b746a14 does sort the keys of the toctree includes -- but it doesn't sort the values (the parent document list).

Note that the `toctree_includes` is a parent -> children mapping. Essentially by sorting the keys (which are parent names) and later iterating over the dictionary, it is guaranteed that parents are processed in lexicographical order. Therefore when the `toctree_includes` is reversed to obtain `toc_parents` i.e child -> parents mapping, the parents list for each child key would already be sorted.

Reply:
> Note that the `toctree_includes` is a parent -> children mapping. Essentially by sorting the keys (which are parent names) and later iterating over the dictionary, it is guaranteed that parents are processed in lexicographical order. Therefore when the `toctree_includes` is reversed to obtain `toc_parents` i.e child -> parents mapping in [check_toc_parents](https://github.com/sphinx-doc/sphinx/commit/8351936994df62aadcc299af3c8106893b746a14#diff-98101714dd35d78a9bdbb8c487fc2e8d10eb3a54772b8da767f34941db21a39dR828-R830), the parents list for each child key would already be sorted.

Ok, this makes sense, thank you.  I didn't understand this until I began attempting to write some unit tests locally.  What I'd started on was a test to provide three or four permutations of the same graph -- sometimes with child value lists randomized, sometimes with parent keys randomized -- with the intent of asserting on consistent inverted-graph structure as output.

(in particular, refreshing my memory about the lexicographic sorting took me some time -- and now I understand from that plus your message here, that the `check` code relies on the parent-key order, and that without a larger refactoring, it wouldn't be valid to write test cases that randomize that ordering, because the application code itself shouldn't allow that to occur)

Reply:
> that the check code relies on the parent-key order, and that without a larger refactoring, it wouldn't be valid to write test cases that randomize that ordering, because the application code itself shouldn't allow that to occur

While it is true that the order of nodes visited during the reversal of the graph edges rely on the parent-key order from `toctree_includes` which is ensured to be deterministic from my previous related PR. However I cannot understand the motivation behind testing the randomized ordering, because in formal definitions a graph consists of **set** of edge pairs, this unordering implies that any permutation of parents lists in `toc_parents` values would correspond to the same graph by definition although different internal representations.  

Was your idea about testing whether the `toc_parents` dictionary order obtained from the dfs is equivalent for any randomized order of key values of a specific graph of `toctree_includes`?